### PR TITLE
Remove entire first line when JSON XSSI prevention seen

### LINKED ISSUE #278: Strip )]} from JSON before parsing
The source map spec allows prepending ")]}" to avoid XSSI attacks. It says:

> Thus when delivering source maps over HTTP, servers may prepend a line starting with the string “)]}'” to the sourcemap. If the response starts with this string clients must ignore the first line.

It's not clear how important this is to support, but if it is important, it would be nice if `SourceMapConsumer` handled this.

### PR DESCRIPTION
**Pull request by tromey**
The source map specification says that the entire first line should be
removed when the source map string starts with the JSON XSSI preventing
string.
Fixes #278

### CODE REVIEW

File: lib/util.js [hunk 1]
```diff
@@ -423,3 +423,21 @@ function compareByGeneratedPositionsInflated(mappingA, mappingB) {
...
 }
 exports.compareByGeneratedPositionsInflated = compareByGeneratedPositionsInflated;
+
+/**
+ * Strip any JSON XSSI avoidance prefix from the string (as documented
+ * in the source maps specification), and then parse the string as
+ * JSON.
+ */
+function parseJSON(str) {
+  if (str.startsWith(")]}'")) {
```
Comment (julienw) [inline lib/util.js]: `startsWith` fails in node 0.10/0.12, sadly: https://travis-ci.org/mozilla/source-map/builds/279505665

Comment (julienw) [reply lib/util.js]: I wouldn't mind to stop supporting these versions but then I think you'll need a major version change.

Comment (julienw) [reply lib/util.js]: I'd reuse the previous regexp to do this as a replacement.

File: lib/source-map-consumer.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ var quickSort = require('./quick-sort').quickSort;
 function SourceMapConsumer(aSourceMap) {
   var sourceMap = aSourceMap;
   if (typeof aSourceMap === 'string') {
-    sourceMap = JSON.parse(aSourceMap.replace(/^\)\]\}'/, ''));
```
Comment (julienw) [inline lib/source-map-consumer.js]: note: if you reuse the same regexp (see below), please note that only the parens needs escaping.

File: lib/util.js [hunk 3]
```diff
@@ -423,3 +423,21 @@ function compareByGeneratedPositionsInflated(mappingA, mappingB) {
...
+    var index = str.indexOf("\n");
+    if (index < 0) {
+      // No line, return nothing.
+      return null;
+    }
+    str = str.substring(index + 1);
```
Comment (julienw) [inline lib/util.js]: Wondering if we should then assert that the source map format revision is > 3. I think we don't need to, but it's your call.

File: lib/util.js [hunk 4]
```diff
@@ -423,3 +423,21 @@ function compareByGeneratedPositionsInflated(mappingA, mappingB) {
...
+ * Strip any JSON XSSI avoidance prefix from the string (as documented
+ * in the source maps specification), and then parse the string as
+ * JSON.
+ */
+function parseJSON(str) {
+  if (str.startsWith(")]}'")) {
+    var index = str.indexOf("\n");
+    if (index < 0) {
+      // No line, return nothing.
+      return null;
```
Comment (julienw) [inline lib/util.js]: I'd rather throw an error, just like JSON.parse throws an error if there is no input.

File: lib/util.js [hunk 5]
```diff
@@ -423,3 +423,21 @@ function compareByGeneratedPositionsInflated(mappingA, mappingB) {
...
+    }
+    str = str.substring(index + 1);
+  }
+  return JSON.parse(str);
+}
+exports.parseJSON = parseJSON;
```
Comment (julienw) [inline lib/util.js]: nitpicking: I don't like the name as it looks like the only thing it does is parsing JSON :) Maybe "parseSourceMapInput" ?

File: lib/util.js (hunk 1 above)
Comment (tromey) [reply lib/util.js]: I'm planning a major version bump anyhow, due to other pending PRs.

0.10 was still seeing new releases last year; so, ugh, I guess we carry on.

File: lib/util.js (hunk 3 above)
Comment (tromey) [reply lib/util.js]: I think that's done elsewhere.

File: lib/source-map-consumer.js (hunk 2 above)
Comment (tromey) [reply lib/source-map-consumer.js]: I did do it this way fwiw.

File: lib/util.js [hunk 6]
```diff
@@ -423,3 +423,13 @@ function compareByGeneratedPositionsInflated(mappingA, mappingB) {
...
 }
 exports.compareByGeneratedPositionsInflated = compareByGeneratedPositionsInflated;
+
+/**
+ * Strip any JSON XSSI avoidance prefix from the string (as documented
+ * in the source maps specification), and then parse the string as
+ * JSON.
+ */
+function parseSourceMapInput(str) {
+  return JSON.parse(str.replace(/^\)]}'(?:[^\n]*?\n)/, ''));
```
Comment (julienw) [inline lib/util.js]: Some simplifications:
* You don't need the grouping :)
* you don't need the ungreedy variant `*?` thanks to your character class

so I think `/^\)]}'[^\n]*\n/` works fine and is simpler :)

Comment (tromey) [reply lib/util.js]: Yes, duh.  Thanks.

**Pull request merged**